### DISCUSSION TITLE
How do you get the quote for a swap?

### DISCUSSION THREAD

**Discussion by kshitijc1506** [aptos-labs/aptos-developer-discussions#111] [Questions] [answered]
Asked by `sumibisht` on Discord (#dev-discussion)

```
const payload = {
    function: "0xc7efb4076dbe143cbcd98cfaaa929ecfc8f299203dfff63b95ccb6bfe19850fa::router::get_amount_in",
    typeArguments: [parseTypeTag(APTOS_COIN), parseTypeTag(DEVNET_BNB)],
    functionArguments: [new U64(APTOS_COIN_AMOUNT)],
  };
const amount_in = (await aptos.view({ payload }));
console.log(amount_in);
```

> But it says it's not a view function, if it's not a view function then how do people get the quote for swap, and its a public function still I can't call it?

**Comment by kshitijc1506** [comment, marked as answer]
It isn't a view function, view functions have the `#[view]` attribute above them. 

Also, you cannot see the return values of functions that are not view. So even if you pay gas and calculate spot price on-chain (`swap_utils` module), there is no way to access the return values via SDK/API/Indexer based on the design of the given Pancakeswap modules. This is because the information is not stored in a resource, nor is it emitted through an event. 

There are multiple ways you can get a quote for a swap, here are just some examples. This answer uses the Pancakeswap modules on [testnet](https://explorer.aptoslabs.com/account/0xc7efb4076dbe143cbcd98cfaaa929ecfc8f299203dfff63b95ccb6bfe19850fa/modules/code/math?network=testnet) as reference. 

1. Since you're using Pancakeswap, you can check out their [aptos-swap-sdk](https://github.com/pancakeswap/pancake-frontend/tree/develop/packages/aptos-swap-sdk). This will likely have a method to get a quote. 


2. Calculate spot price by using AMM math based on token reserves, you can do this yourself off-chain. This can be more efficient and is often used by DEX aggregators for larger scale applications. You can get the reserves by reading the fields of the `TokenPairReserve` resource on the `swap` module (or you can keep track of this information off-chain for speed). This can be done via API, [SDK](https://aptos.dev/sdks/ts-sdk/fetch-data-from-chain#queries-with-generics), or indexer. 

API endpoint format is below:
`http://<API_ENDPOINT>/v1/accounts/<ACCOUNT_ADDRESS>/resource/<RESOURCE_TYPE>`

The logic for the math can be found [here](https://explorer.aptoslabs.com/account/0xc7efb4076dbe143cbcd98cfaaa929ecfc8f299203dfff63b95ccb6bfe19850fa/modules/code/swap_utils?network=testnet). Look at `get_amount_in` and `get_amount_out` functions. Here's a more detailed explanation of the [math](https://www.youtube.com/watch?v=a56XeddkOtA).


3. Use the TypeScript SDK to [simulate](https://aptos.dev/sdks/ts-sdk/transaction-builder/#simulate-transaction) the swap. 

Here's a code snippet that demonstrates swap simulation:
```
// create a new instance of the Aptos SDK
  const config = new AptosConfig({ network: Network.TESTNET }); 
  const aptos = new Aptos(config);

  // generate and fund new account
  const alice: Account = Account.generate();
  await aptos.fundAccount({
    accountAddress: alice.accountAddress,
    amount: 200000000,
  });

    // build the transaction
  const transaction = await aptos.transaction.build.simple({
    sender: alice.accountAddress,
    data: {
      function:
        "0xc7efb4076dbe143cbcd98cfaaa929ecfc8f299203dfff63b95ccb6bfe19850fa::router::swap_exact_input",
      typeArguments: ["0x1::aptos_coin::AptosCoin", "0x16fe2df00ea7dde4a63409201f7f4e536bde7bb7335526a35d05111e68aa322c::TestCoinsV1::USDT"],
      functionArguments: [10000000, 0],
    },
  });

  // simulate the transaction
  const [userTransactionResponse] = await aptos.transaction.simulate.simple({
    signerPublicKey: alice.publicKey,
    transaction,
  });
  console.log(userTransactionResponse.events);
```

**Reply by degenSumi** [reply]
Thanks for answering will check it out,
One question though on swap_utils module I see amout_out function however I still don't see any #view attribute?

**Reply by kshitijc1506** [reply]
I edited the answer, let me know if it makes sense!

**Comment by degenSumi** [comment]
Thanks @kshitjee it helped.

**Comment by degenSumi** [comment]
Hey @kshitjee just a quick question, now how can i send this transaction 

const txn = await aptos.transaction.build.simple({
    sender: "0xbe5cd08d8fe346b9b55dba64fc09e50dda92a8bd16743652ec8b156821123310",
    data: {
      function: "0xc7efb4076dbe143cbcd98cfaaa929ecfc8f299203dfff63b95ccb6bfe19850fa::router::swap_exact_input",
      typeArguments: [parseTypeTag(APTOS_COIN), parseTypeTag(DEVNET_BNB)],
      functionArguments: [new U64(APTOS_COIN_AMOUNT), new U64(5)],
    },
  });
  console.log(txn);
  
  // Now i want to encode this tx and send it over the internet so that anyone else can sign it

  const signedTx = aptos.transaction.sign({
    signer:signer,
    transaction: txn
  });

// Now i want to send this signed tx so that any one else can send this 
  console.log(signedTx.deserialize(), txn);
  
  please let me know as i can't find any functions for it on the sdk
